File: src/oca/go/src/goca/host.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2023, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package goca

import (
	"context"
	"encoding/xml"
	"errors"

	"github.com/OpenNebula/one/src/oca/go/src/goca/parameters"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/host"
)

// HostsController is a controller for create an host or a pool of hosts
type HostsController entitiesController

// HostController is a controller for host entities
type HostController entityController

// Hosts returns a hosts controller.
func (c *Controller) Hosts() *HostsController {
	return &HostsController{c}
}

// Host return an host controller with an ID.
func (c *Controller) Host(id int) *HostController {
	return &HostController{c, id}
}

// ByName finds a Host ID from name
func (c *HostsController) ByName(name string) (int, error) {
	return c.ByNameContext(context.Background(), name)
}

// ByNameContext finds a Host ID from name
func (c *HostsController) ByNameContext(ctx context.Context, name string) (int, error) {
	var id int

	hostPool, err := c.InfoContext(ctx)
	if err != nil {
		return -1, err
	}

	match := false
	for i := 0; i < len(hostPool.Hosts); i++ {
		if hostPool.Hosts[i].Name != name {
			continue
		}
		if match {
			return -1, errors.New("multiple resources with that name")
		}
		id = hostPool.Hosts[i].ID
		match = true
	}
	if !match {
		return -1, errors.New("resource not found")
	}
	return id, nil
}

// Info returns a host pool. A connection to OpenNebula is
// performed
func (hc *HostsController) Info() (*host.Pool, error) {
	return hc.InfoContext(context.Background())
}

// InfoContext returns a host pool. A connection to OpenNebula is
// performed
func (hc *HostsController) InfoContext(ctx context.Context) (*host.Pool, error) {
	response, err := hc.c.Client.CallContext(ctx, "one.hostpool.info")
	if err != nil {
		return nil, err
	}

	hostPool := &host.Pool{}
	err = xml.Unmarshal([]byte(response.Body()), &hostPool)
	if err != nil {
		return nil, err
	}
	return hostPool, nil
}

// Info retrieves information for the host from ID
func (hc *HostController) Info(decrypt bool) (*host.Host, error) {
	return hc.InfoContext(context.Background(), decrypt)
}

// InfoContext retrieves information for the host from ID
func (hc *HostController) InfoContext(ctx context.Context, decrypt bool) (*host.Host, error) {
	response, err := hc.c.Client.CallContext(ctx, "one.host.info", hc.ID, decrypt)
	if err != nil {
		return nil, err
	}
	host := &host.Host{}
	err = xml.Unmarshal([]byte(response.Body()), host)
	if err != nil {
		return nil, err
	}
	return host, nil
}

// Create allocates a new host. It returns the new host ID.
// * name: name of the host
// * im: information driver for the host
// * vm: virtualization driver for the host
// * clusterID: The cluster ID. If it is -1, the default one will be used.
func (hc *HostsController) Create(name, im, vm string, clusterID int) (int, error) {
	return hc.CreateContext(context.Background(), name, im, vm, clusterID)
}

// Create allocates a new host. It returns the new host ID.
// * ctx: context for cancelation
// * name: name of the host
// * im: information driver for the host
// * vm: virtualization driver for the host
// * clusterID: The cluster ID. If it is -1, the default one will be used.
func (hc *HostsController) CreateContext(ctx context.Context, name, im, vm string, clusterID int) (int, error) {
	response, err := hc.c.Client.CallContext(ctx, "one.host.allocate", name, im, vm, clusterID)
	if err != nil {
		return -1, err
	}

	return response.BodyInt(), nil
}

// Monitoring Returns the Hosts monitoring records
// num: Retrieve monitor records in the last num seconds.
// 0 just the last record, -1 all records
func (hc *HostsController) Monitoring(num int) (*host.PoolMonitoring, error) {
	return hc.MonitoringContext(context.Background(), num)
}

// MonitoringContext Returns the Hosts monitoring records
// ctx: context for cancelation
// num: Retrieve monitor records in the last num seconds.
// 0 just the last record, -1 all records
func (hc *HostsController) MonitoringContext(ctx context.Context, num int) (*host.PoolMonitoring, error) {
	monitorData, err := hc.c.Client.CallContext(ctx, "one.hostpool.monitoring", num)
	if err != nil {
		return nil, err
	}

	hostsMon := &host.PoolMonitoring{}
	err = xml.Unmarshal([]byte(monitorData.Body()), &hostsMon)
	if err != nil {
		return nil, err
	}

	return hostsMon, nil
}

// Delete deletes the given host from the pool
func (hc *HostController) Delete() error {
	return hc.DeleteContext(context.Background())
}

// DeleteContext deletes the given host from the pool
func (hc *HostController) DeleteContext(ctx context.Context) error {
	_, err := hc.c.Client.CallContext(ctx, "one.host.delete", hc.ID)
	return err
}

// Status sets the status of the host
// * status: 0: ENABLED, 1: DISABLED, 2: OFFLINE
func (hc *HostController) Status(status int) error {
	return hc.StatusContext(context.Background(), status)
}

// StatusContext sets the status of the host
// * ctx: context for cancelation
// * status: 0: ENABLED, 1: DISABLED, 2: OFFLINE
func (hc *HostController) StatusContext(ctx context.Context, status int) error {
	_, err := hc.c.Client.CallContext(ctx, "one.host.status", hc.ID, status)
	return err
}

// Update adds host content.
//   - tpl: The new host contents. Syntax can be the usual attribute=value or XML.
//   - uType: Update type: Replace: Replace the whole template.
//     Merge: Merge new template with the existing one.
func (hc *HostController) Update(tpl string, uType parameters.UpdateType) error {
	return hc.UpdateContext(context.Background(), tpl, uType)
}

// UpdateContext adds host content.
//   - ctx: context for cancelation
//   - tpl: The new host contents. Syntax can be the usual attribute=value or XML.
//   - uType: Update type: Replace: Replace the whole template.
//     Merge: Merge new template with the existing one.
func (hc *HostController) UpdateContext(ctx context.Context, tpl string, uType parameters.UpdateType) error {
	_, err := hc.c.Client.CallContext(ctx, "one.host.update", hc.ID, tpl, uType)
	return err
}

// Rename renames a host.
// * newName: The new name.
func (hc *HostController) Rename(newName string) error {
	return hc.RenameContext(context.Background(), newName)
}

// RenameContext renames a host.
// * ctx: context for cancelation
// * newName: The new name.
func (hc *HostController) RenameContext(ctx context.Context, newName string) error {
	_, err := hc.c.Client.CallContext(ctx, "one.host.rename", hc.ID, newName)
	return err
}

// Monitoring returns the host monitoring records.
func (hc *HostController) Monitoring() (*host.Monitoring, error) {
	return hc.MonitoringContext(context.Background())
}

// MonitoringContext returns the host monitoring records.
func (hc *HostController) MonitoringContext(ctx context.Context) (*host.Monitoring, error) {
	monitorData, err := hc.c.Client.CallContext(ctx, "one.host.monitoring", hc.ID)
	if err != nil {
		return nil, err
	}

	hostMon := &host.Monitoring{}
	err = xml.Unmarshal([]byte(monitorData.Body()), &hostMon)
	if err != nil {
		return nil, err
	}

	return hostMon, nil
}

File: src/fireedge/src/client/features/OneApi/host.js
/* ------------------------------------------------------------------------- *
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems               *
 *                                                                           *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may   *
 * not use this file except in compliance with the License. You may obtain   *
 * a copy of the License at                                                  *
 *                                                                           *
 * http://www.apache.org/licenses/LICENSE-2.0                                *
 *                                                                           *
 * Unless required by applicable law or agreed to in writing, software       *
 * distributed under the License is distributed on an "AS IS" BASIS,         *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
 * See the License for the specific language governing permissions and       *
 * limitations under the License.                                            *
 * ------------------------------------------------------------------------- */
import { Actions, Commands } from 'server/utils/constants/commands/host'

import { Host } from 'client/constants'
import {
  ONE_RESOURCES,
  ONE_RESOURCES_POOL,
  oneApi,
} from 'client/features/OneApi'
import {
  removeResourceOnPool,
  updateNameOnResource,
  updateResourceOnPool,
  updateTemplateOnResource,
} from 'client/features/OneApi/common'

import {
  Actions as ExtraActions,
  Commands as ExtraCommands,
} from 'server/routes/api/host/routes'

import { UpdateFromSocket } from 'client/features/OneApi/socket'

const { HOST } = ONE_RESOURCES
const { HOST_POOL } = ONE_RESOURCES_POOL

const hostApi = oneApi.injectEndpoints({
  endpoints: (builder) => ({
    getHosts: builder.query({
      /**
       * Retrieves information for all the hosts in the pool.
       *
       * @param {object} params - Request params
       * @param {string} [params.zone] - Zone from where to get the resources
       * @returns {Host[]} Get list of hosts
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_POOL_INFO
        const command = { name, ...Commands[name] }

        return { command, params }
      },
      transformResponse: (data) => [data?.HOST_POOL?.HOST ?? []].flat(),
      providesTags: (hosts) =>
        hosts
          ? [
              ...hosts.map(({ ID }) => ({ type: HOST_POOL, id: `${ID}` })),
              HOST_POOL,
            ]
          : [HOST_POOL],
    }),
    getHostsAdmin: builder.query({
      /**
       * Retrieve the information as serveradmin.
       *
       * @param {object} params - Request params
       * @returns {Host[]} Get cluster identified by id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = ExtraActions.HOSTPOOL_ADMINSHOW
        const command = { name, ...ExtraCommands[name] }

        return { params, command }
      },
      transformResponse: (data) => [data?.HOST_POOL?.HOST ?? []].flat(),
      providesTags: (hosts) =>
        hosts
          ? [
              ...hosts.map(({ ID }) => ({ type: HOST_POOL, id: `${ID}` })),
              HOST_POOL,
            ]
          : [HOST_POOL],
    }),
    getHost: builder.query({
      /**
       * Retrieves information for the host.
       *
       * @param {object} params - Request params
       * @param {string} params.id - Host id
       * @returns {Host} Get host identified by id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => {
        if (!data?.HOST) return {}

        const monitoring = data?.HOST?.MONITORING

        const hostShare = data?.HOST?.HOST_SHARE?.NUMA_NODES

        if (!monitoring || !hostShare) return data.HOST

        /**
         * [GH-6027] Numa nodes attributes are not together, some of the attributes
         * now come from monitoring. This means that when we ask for the host
         * data we need to merge the numa monitoring data with the numa host
         * share data
         */
        const numaNodes =
          hostShare?.NODE && Array.isArray(hostShare.NODE)
            ? hostShare.NODE
            : [hostShare.NODE]

        const monitoringNodes = Array.isArray(monitoring.NUMA_NODE)
          ? monitoring.NUMA_NODE
          : [monitoring.NUMA_NODE]

        numaNodes.forEach((node) => {
          const monitoringNode = monitoringNodes.find(
            (mNode) => mNode?.NODE_ID === node?.NODE_ID
          )
          node.MEMORY.FREE = monitoringNode.MEMORY.FREE
          node.MEMORY.USED = monitoringNode.MEMORY.USED

          const ensuredHugepage =
            node.HUGEPAGE && !Array.isArray(node.HUGEPAGE)
              ? [node.HUGEPAGE]
              : node.HUGEPAGE

          ensuredHugepage.forEach((page) => {
            const ensuredMonitoringNodeHugepage =
              monitoringNode.HUGEPAGE && !Array.isArray(monitoringNode.HUGEPAGE)
                ? [monitoringNode.HUGEPAGE]
                : monitoringNode.HUGEPAGE
            const monitoringPage = ensuredMonitoringNodeHugepage.find(
              (mPage) => mPage.SIZE === page.SIZE
            )
            page.FREE = monitoringPage.FREE

            return page
          })

          return node
        })

        return data.HOST
      },
      providesTags: (_, __, { id }) => [{ type: HOST, id }],
      async onQueryStarted({ id }, { dispatch, queryFulfilled }) {
        try {
          const { data: resourceFromQuery } = await queryFulfilled

          dispatch(
            hostApi.util.updateQueryData(
              'getHosts',
              undefined,
              updateResourceOnPool({ id, resourceFromQuery })
            )
          )
        } catch {
          // if the query fails, we want to remove the resource from the pool
          dispatch(
            hostApi.util.updateQueryData(
              'getHosts',
              undefined,
              removeResourceOnPool({ id })
            )
          )
        }
      },
      onCacheEntryAdded: UpdateFromSocket({
        updateQueryData: (updateFn) =>
          hostApi.util.updateQueryData('getHosts', undefined, updateFn),
        resource: 'HOST',
      }),
    }),
    allocateHost: builder.mutation({
      /**
       * Allocates a new host in OpenNebula.
       *
       * @param {object} params - Request params
       * @param {string} params.hostname - Hostname of the machine we want to add
       * @param {string} params.imMad
       * - The name of the information manager (im_mad_name),
       * this values are taken from the oned.conf with the tag name IM_MAD (name)
       * @param {string} params.vmmMad
       * - The name of the virtual machine manager mad name (vmm_mad_name),
       * this values are taken from the oned.conf with the tag name VM_MAD (name)
       * @param {string|number} [params.cluster] - The cluster ID
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_ALLOCATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
    }),
    updateHost: builder.mutation({
      /**
       * Replaces the host’s template contents.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - Host id
       * @param {string} params.template - The new template contents
       * @param {0|1} params.replace
       * - Update type:
       * ``0``: Replace the whole template.
       * ``1``: Merge new template with the existing one.
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_UPDATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: HOST, id }],
      async onQueryStarted(params, { dispatch, queryFulfilled }) {
        try {
          const patchHost = dispatch(
            hostApi.util.updateQueryData(
              'getHost',
              { id: params.id },
              updateTemplateOnResource(params)
            )
          )

          const patchHosts = dispatch(
            hostApi.util.updateQueryData(
              'getHosts',
              undefined,
              updateTemplateOnResource(params)
            )
          )

          queryFulfilled.catch(() => {
            patchHost.undo()
            patchHosts.undo()
          })
        } catch {}
      },
    }),
    removeHost: builder.mutation({
      /**
       * Deletes the given host from the pool.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - Host id
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_DELETE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [HOST_POOL],
    }),
    enableHost: builder.mutation({
      /**
       * Sets the status of the host to enabled.
       *
       * @param {number|string} id - Host id
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (id) => {
        const name = Actions.HOST_STATUS
        const command = { name, ...Commands[name] }

        return { params: { id, status: 0 }, command }
      },
      invalidatesTags: (_, __, id) => [{ type: HOST, id }, HOST_POOL],
    }),
    disableHost: builder.mutation({
      /**
       * Sets the status of the host to disabled.
       *
       * @param {number|string} id - Host id
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (id) => {
        const name = Actions.HOST_STATUS
        const command = { name, ...Commands[name] }

        return { params: { id, status: 1 }, command }
      },
      invalidatesTags: (_, __, id) => [{ type: HOST, id }, HOST_POOL],
    }),
    offlineHost: builder.mutation({
      /**
       * Sets the status of the host to offline.
       *
       * @param {number|string} id - Host id
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (id) => {
        const name = Actions.HOST_STATUS
        const command = { name, ...Commands[name] }

        return { params: { id, status: 2 }, command }
      },
      invalidatesTags: (_, __, id) => [{ type: HOST, id }, HOST_POOL],
    }),
    renameHost: builder.mutation({
      /**
       * Renames a host.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Host id
       * @param {string} params.name - New name
       * @returns {number} Host id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_RENAME
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: HOST, id }],
      async onQueryStarted(params, { dispatch, queryFulfilled }) {
        try {
          const patchHost = dispatch(
            hostApi.util.updateQueryData(
              'getHost',
              { id: params.id },
              updateNameOnResource(params)
            )
          )

          const patchHosts = dispatch(
            hostApi.util.updateQueryData(
              'getHosts',
              undefined,
              updateNameOnResource(params)
            )
          )

          queryFulfilled.catch(() => {
            patchHost.undo()
            patchHosts.undo()
          })
        } catch {}
      },
    }),
    getHostMonitoring: builder.query({
      /**
       * Returns the host monitoring records.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Host id
       * @returns {string} The monitoring information string / The error string
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_MONITORING
        const command = { name, ...Commands[name] }

        return { params, command }
      },
    }),
    getHostMonitoringPool: builder.query({
      /**
       * Returns all the host monitoring records.
       *
       * @param {object} params - Request parameters
       * @param {number|'0'|'-1'} [params.seconds]
       * - Retrieve monitor records in the last num seconds.
       * `0`: Only the last record.
       * `-1`: All records.
       * @returns {string} The monitoring information string / The error string
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.HOST_POOL_MONITORING
        const command = { name, ...Commands[name] }

        return { params, command }
      },
    }),
  }),
})

export const {
  // Queries
  useGetHostQuery,
  useLazyGetHostQuery,
  useGetHostsQuery,
  useGetHostsAdminQuery,
  useLazyGetHostsQuery,
  useGetHostMonitoringQuery,
  useLazyGetHostMonitoringQuery,
  useGetHostMonitoringPoolQuery,
  useLazyGetHostMonitoringPoolQuery,

  // Mutations
  useAllocateHostMutation,
  useUpdateHostMutation,
  useRemoveHostMutation,
  useEnableHostMutation,
  useDisableHostMutation,
  useOfflineHostMutation,
  useRenameHostMutation,
} = hostApi

export default hostApi
